### DISCUSSION TITLE
Possibility for kustomize-controller to fail early in health checks to accommodate inherently slow resource types

### DISCUSSION THREAD

**Discussion by Xitric** [fluxcd/flux2#3980] [Proposals]
## Problem

To be clear, I have been digging around the discussions and issues for [flux2](https://github.com/fluxcd/flux2) and [kustomize-controller](https://github.com/fluxcd/kustomize-controller) to understand the behaviour of Kustomization health checks and their limitations. As I understand, the general issue is that a Kustomization health checks rely on the behaviour of the controller-runtime queue, meaning that once a health check is queued with a timeout of say 10 minutes, a failure cannot be reported until the 10 minutes have passed (https://github.com/fluxcd/kustomize-controller/issues/487#issuecomment-966487391) - even if the resource being reconciled could not possibly recover on its own.

I also know that the simple solution is to set a smaller timeout so that we won't need to wait as long (https://github.com/fluxcd/flux2/issues/1577#issuecomment-890898999).

Still, I believe that there are genuine cases where it could be beneficial to support early-out failure of Kustomizations - that is, failure reporting before the health checks time out. I will get back to this shortly. First, I wish to outline what I hope to gather from this discussion:

1. **Is there any possibility of supporting early failure for Kustomizations?**
    1. **if so, what would be required of Flux2 and others?**
    1. **If not, what are the recommendations for workarounds or alternatives? (keeping in mind the cases I outline below)**

I know that this is a topic that has been brought up multiple times in the Flux2 community, and still some issues remain open - I hope you are not tired of me bringing this up once more, but I would like to have a more focused discussion this time around:

- #1577
- #1000
- #2816
- #2987
- #2989
- #3725
- fluxcd/kustomize-controller#701
- fluxcd/kustomize-controller#487
- fluxcd/kustomize-controller#405

## Use case 1: Crossplane resources

At my company, we have started to look into using tools such as CrossPlane or Azure Service Operator to manage cloud services via Kubernetes resources. However, my experience tells me that certain such cloud resources do not tend to deploy very fast. Managed Redis caches in Azure often take 15-20 minutes to provision, and VPN gateways can take more than an hour. These are just a few examples.

If I were to reconcile such resources using Flux and I depend on ordering deployments (say, provision cloud services with one Kustomization before deploying the application pods using another Kustomization), I would need to set timeouts of more than an hour to avoid errors. But if a deployment fails early on, we cannot recover until after the timeout unless we manually connect to the otherwise hands-off Kubernetes cluster and start killing kustomize-controller pods (fluxcd/kustomize-controller#405).

## Use case 2: Long running database migrations

At my company, we love GitOps and we try to abide to GitOps principles in everything that we do. Often when we upgrade our applications in Kubernetes, we need to perform database migrations as part of the rollout. In order to do this declaratively, we have opted for both:

- Using Helm install, pre-upgrade, and post-upgrade hooks
- Using dependent Flux Kustomizations for running pre- and post-deployment jobs (https://fluxcd.io/flux/use-cases/running-jobs/)

Our applications, however, are not simple, and the databases often contain hundreds of millions of rows that could be in need of changes during such a migration. This means that a migration from one version to the next could take hours to complete in the worst case, but often it is much faster. We have hundreds of such databases, and while we _could_ trigger the migration out-of-band prior to rolling out a new version, that causes obstacles in our rollout processes and goes against the principles of GitOps. Suddenly, rolling out a new application version is no longer just an act of merging a pull request to main. We need to develop a separate, coordinated process for migrating databases and only then triggering application rollouts. We hope to rely on GitOps for this as well.

One alternative of course is to split up the migration into multiple, shorter steps. In fact, we do this already inside the migration job. A very long migration is handled as a series of checkpointed operations, because we wish to support the case when Kubernetes or hardware failures cause a job to end prematurely. Our migration jobs are guaranteed to be idempotent and retryable - they just pick up from the checkpoint where they left off. But each step could still take >15 minutes.

In order to make these steps in the migrations visible to Flux, we would need to create a chain of interdependent Kustomizations - one for each step in the migration. This changes for every application version. Now we are stating to make our configuration imperative!  It seems much better if we could simply state: "Hey job, just make sure that the database is up to date", and let all concerns about individual migration steps be abstracted away from Flux and be in the hands of the developer.

To prevent Flux from givng up on a migration midway, we need to set a long timeout on our pre- and post-deployment Kustomizations in Flux, or on the HelmRelease resource when using Helm hooks. If anything fails early in the migration process, we cannot recover until after the timeout unless we manually connect to the cluster once more.

## Epilogue

So, I really wish to get some thoughts on these challenges. I completely understand if early failures of Kustomizations is not a trivial thing in Flux, and I am also very open to looking at the problem from different angles! But currently, this limitation concerns me somewhat.

I did take a look at the source code though, and I can see that a lot of the status checking logic in Flux is at least two years old: https://github.com/fluxcd/flux2/blob/main/pkg/status/status.go
However, in November 2021 kstatus implemented support for handling failed resources (avoiding waiting for the timeout): kubernetes-sigs/cli-utils#486. I wonder if Flux simply needs to add support for this new Failed event?

Thank you for a great product though!

**Comment by Xitric** [comment]
@kingdonb @stefanprodan Would it be viable to support the new events from https://github.com/kubernetes-sigs/cli-utils/pull/486 in Flux2?

**Reply by somtochiama** [reply]
Hey @Xitric ,

Thanks for the detailed explanation. The pull request you have linked is to the `apply` pkg in cli-utils, but in flux, we use the [kstatus](https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/README.md) pkg for health checking resources.

However, I think some parts of what the pull request does might still be relevant to the issue, which is to stop waiting on resources to be ready if there are `stalled` resources AND the rest of them are ready.  Especially for resources that the controller is managing (i.e. when you set `spec.wait` to true) since it most likely needs an update to get out of the stalled state and the controller is the one that's going to apply that update. So it makes sense for the controller to return before the timeout out and retry. For the specific health checks defined on the Kustomization, I am less sure. The Kustomization might not be what's managing the resource and something else can come in and fix it before the timeout is done.

I think we can do this by updating the [observerFunc](https://github.com/fluxcd/pkg/blob/5b608cb95e2129ad9cf19ed6e26f87003518a70e/ssa/manager_wait.go#L81)  to check for FailedStatus from kstatus :). 
> Failed: The process of reconciling the actual state with the desired state has encountered and error or it has made insufficient progress.

However, I think this would require more discussion w other maintainers (c @stefanprodan @darkowlzz )

**Reply by stefanprodan** [reply]
@somtochiama let’s give this a try, I think we need to make it optional. Not sure how stalled is determined for things like Deployments and Statefulsets, if lets say, it encounters an image pull backoff is it stalled? most of these errors can be transient, should we bail out early even if Kubernetes doesn’t?

**Reply by Xitric** [reply]
Thanks both of you for the reply! I didn't even notice that the pull request was related to a different tool than kstatus, but I am happy to hear that we might still be able to make some headway on this.

> I think we need to make it optional. Not sure how stalled is determined for things like Deployments and Statefulsets, if lets say, it encounters an image pull backoff is it stalled?

In terms of making it optional, do you mean in the sense that based on our domain knowledge, we would be able to define whether we consider all/some transient errors such as image pull backoff to be a stall? Or rather that the whole concept of detecting stalled resources should be optional (and then we need to more carefully consider what is truly stalled)? I am certainly open for both.

I can give some examples of what we would benefit the most from detecting:
- Determining that a HelmRelease is stalled if we receive "install retries exhausted"
- Determining that a Job is stalled if it reaches its `backoffLimit`

In both of these cases we can configure limits to retries in Kubernetes, which makes it possible for us to define when we believe a possibly transient error will no longer be able to recover.

For other cases, like determining whether a deployment is stalled when it keeps getting restarted by the kubelet, I acknowledge that it is probably not possible to generalize. But we are open to any solution that brings us even a small step closer to early detection of stalls.

**Reply by stefanprodan** [reply]
It’s kstatus logic that determines when something is stalled, I suspect the stalled condition is rarely retuned. kstatus doesn’t understands jobs at all, this is something that we had to handcraft in Flux so we may be able to fail fast for those but that’s the exception.

**Reply by somtochiama** [reply]
Yes, like @stefanprodan said we are tied to the kstatus logic atm, the logic for some core types are defined here https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/status/core.go.
For Flux types, we follow the kstatus recommendations and place the `Stalled` conditions on the object.
For other Kubernetes types, it isn't as straightforward

>  Deployments and Statefulsets, if let's say, it encounters an image pull backoff is it stalled?

For Deployments, kstatus only returns a FailedStatus(the equivalent of stalled for kstatus) when the `ProgressDeadlineExceeded` condition is set. If the pods aren't able to come out of image pullback off before the ProgressDeadlineSeconds`, the deployment is marked as Failed.

For Pods(i.e. the user specified a Pod in the health check, not a deployment), a failed status is only returned when the pod is stuck `CrashLoopBackOff` or unschedulable. So for an image pull back off, an InProgess status is returned for the pod.

I think the worry, is that this might not work as users might anticipate e.g. eventually the image pullback off will cause a FailedStatus for a Deployment but not for a Pod.

**Reply by stefanprodan** [reply]
Currently a HelmRelease doesn’t have a Stalled condition so kstatus can’t fail fast, @hiddeco please correct me if I’m wrong.

**Reply by hiddeco** [reply]
It does not have a `Stalled` condition at present, [but will](https://github.com/fluxcd/helm-controller/blob/dev/internal/reconcile/atomic_release.go#L148) once we have landed the next bit of the roadmap.

**Reply by stefanprodan** [reply]
@somtochiama the usage of `ProgressDeadlineExceeded` is great, we don’t care about pods as Flux shouldn’t manage those directly. As for jobs, the implementation here looks good to me https://github.com/fluxcd/kustomize-controller/blob/main/internal/statusreaders/job.go#L89. Please go ahead and add a fail fast option to `ssa.Wait`.

**Reply by Xitric** [reply]
This all sounds perfect to us, and we are really looking forward to seeing a fail fast option in Flux.

For our use cases, we do not need to manage pods directly either, so this is fine.

We hope that it will be possible to have custom stall detection for jobs, since we rely on those in our infrastructure. But if not, at least the upcoming support for stall detection in HelmReleases will be a big improvement for us.

As for relying on `ProgressDeadlineExceeded` with Deployments, this is also a great fit for us. It will enable us to set a relatively high timeout on kustomizations with `wait: true`, to give ample time for jobs, Helm hooks, and Crossplane resources to finish, but still mange to detect and recover from non-transient Deployment issues in a few minutes (as opposed to, say, half an hour) by setting a lower `.spec.progressDeadlineSeconds`.
